File: src/data/dsh.js
'use strict';

// DSH（DeepSeek Harness）用量统计 — 读 ~/.dsh/sessions 的 zstd 压缩会话日志。
// token 口径对齐 DSH GUI（token-meter/StatsLine）：usage 来自 assistant/chunk 的
// usage 类型 chunk 与 assistant/message 的 data.usage，同 turn/step 只取最后一份
// （后到替换，不累加）；输入 = uncachedInput + cacheRead + cacheWrite（billed 口径，
// GUI 的 billedInputTokens）。会话详情：session/title（标题）、user/message（首条
// 消息）、request/context（模型）、tool/call（工具名）。
// 15s TTL 缓存（{ts,payload} 模式），status() 路径零网络零阻塞。

const fs = require('fs');
const path = require('path');
const os = require('os');
const { zstdDecompressSync } = require('node:zlib');

const SESSIONS_ROOT = path.join(os.homedir(), '.dsh', 'sessions');
const CACHE_TTL = 15.0;
const ZSTD_MAGIC = Buffer.from([0x28, 0xb5, 0x2f, 0xfd]); // zstd frame magic
const MAX_SESSION_LIST = 10; // 会话卡列表只回最近 N 个，聚合仍全量

let cache = { ts: 0, payload: null };

// 逐帧解压 zstd 流（每帧独立 frame，循环找 magic 边界解压拼接）
function decompressZstd(buf) {
  let out = '', pos = 0, frames = 0;
  while (pos < buf.length) {
    const idx = buf.indexOf(ZSTD_MAGIC, pos);
    if (idx < 0) break;
    const next = buf.indexOf(ZSTD_MAGIC, idx + 4);
    const end = next < 0 ? buf.length : next;
    try {
      out += zstdDecompressSync(buf.subarray(idx, end)).toString('utf8');
    } catch (e) {
      break; // 尾帧不完整（正在写入）：丢弃
    }
    pos = end;
    // 坑（2026-08-14 实测）：Electron 35 的 zstdDecompressSync 输出 Buffer 进
    // V8 老生代，JS 堆压力不足时 major GC 永不触发（external 不计数），大量帧
    // 的 Buffer 累积到数 GB → 进程 OOM 无声退出（exit -36861，无事件日志）。
    // 每 200 帧用临时大数组制造堆压力强制 major GC：单文件峰值 1.5GB→390MB。
    if (++frames % 200 === 0) {
      const junk = new Array(1 << 22);
      junk.length = 0;
    }
  }
  return out;
}

// 单步 token 桶。input 是 billed 口径（含缓存），cache 单独拆出供展示。
const ZERO = { input: 0, cache: 0, output: 0, reasoning: 0 };

function addTokens(a, b) {
  a.input += b.input || 0; a.cache += b.cache || 0;
  a.output += b.output || 0; a.reasoning += b.reasoning || 0;
}

// 从 usage 记录取单步 token（输入= billed 口径，对齐 GUI billedInputTokens）
function tokensFromUsage(u) {
  if (!u || typeof u !== 'object') return null;
  const input = u.inputTokens, out = u.outputTokens;
  if (typeof input !== 'number' && typeof out !== 'number') return null;
  const cache = (u.cacheReadTokens || 0) + (u.cacheWriteTokens || 0);
  return {
    input: (typeof input === 'number' ? input : 0) + cache,
    cache,
    output: typeof out === 'number' ? out : 0,
    reasoning: u.reasoningTokens || 0,
  };
}

// 统计单个会话：精确 token + 标题/模型/工具/时间
function statLog(logPath, sid) {
  let text;
  try {
    const buf = fs.readFileSync(logPath);
    text = logPath.endsWith('.zstd') ? decompressZstd(buf) : buf.toString('utf8');
  } catch (e) {
    return null;
  }
  const s = {
    id: sid, title: '', firstMsg: '', model: '', cwd: '',
    turns: 0, steps: 0, toolCalls: 0, tools: {},
    tokens: { ...ZERO }, tokensToday: { ...ZERO }, createdAt: 0, lastTs: 0,
  };
  // 同 turn/step 的 usage 只留最后一份（chunk 先行、message 收尾，后到替换不累加）
  const stepUsage = new Map();
  // 按模型聚合：request/context 声明当前请求模型，其后 usage 归该模型
  let currentModel = '';
  const modelTokens = new Map();
  for (const line of text.split('\n')) {
    let j;
    try { j = JSON.parse(line); } catch (e) { continue; }
    if (!j || typeof j !== 'object') continue;
    const t = j.type, d = j.data || {};
    if (t === 'session') {
      s.createdAt = Number(j.createdAt || 0);
      // 真实 cwd（workspace 目录名是有损编码，不可靠；basename 拿最后一层文件夹）
      if (typeof j.cwd === 'string' && j.cwd) s.cwd = j.cwd;
    } else if (t === 'turn/start' || t === 'turn/end') s.turns++;
    else if (t === 'step/start' || t === 'step/end') s.steps++;
    else if (t === 'tool/call') {
      s.toolCalls++;
      if (d.name) s.tools[d.name] = (s.tools[d.name] || 0) + 1;
    } else if (t === 'request/context') {
      if (d.model || d.provider) {
        s.model = s.model || (d.model || d.provider);
        currentModel = (d.provider ? d.provider + '/' : '') + (d.model || '');
      }
    } else if (t === 'session/title') {
      if (d.title) s.title = d.title;
    } else if (t === 'user/message') {
      const c = d.content;
      if (!s.firstMsg && Array.isArray(c)) {
        const txt = c.find(x => x && x.type === 'text' && typeof x.text === 'string');
        if (txt) s.firstMsg = txt.text;
      }
    } else if (t === 'assistant/chunk') {
      // usage 类型 chunk：流的早期样本（token-meter 同源）
      if (d.chunk && d.chunk.type === 'usage' && d.chunk.usage) {
        stepUsage.set(d.turn + ':' + d.step, { tk: tokensFromUsage(d.chunk.usage), model: currentModel, ts: Number(j.time || 0) });
      }
    } else if (t === 'assistant/message') {
      // 组装消息的最终 usage：覆盖同 step 的 chunk 样本
      if (d.usage) stepUsage.set(d.turn + ':' + d.step, { tk: tokensFromUsage(d.usage), model: currentModel, ts: Number(j.time || 0) });
    }
    const ts = Number(j.time || 0);
    if (ts > s.lastTs) s.lastTs = ts;
  }
  // 今日窗口：本地自然日 0 点（与 sqlite readUsage 同口径）
  const todayStartMs = new Date(new Date().getFullYear(), new Date().getMonth(), new Date().getDate()).getTime();
  const todayModelTokens = new Map();
  for (const { tk, model, ts } of stepUsage.values()) {
    if (!tk) continue;
    addTokens(s.tokens, tk);
    const key = model || 'unknown';
    const bucket = modelTokens.get(key) || { ...ZERO };
    addTokens(bucket, tk);
    modelTokens.set(key, bucket);
    if (ts >= todayStartMs) {
      addTokens(s.tokensToday, tk);
      const tb = todayModelTokens.get(key) || { ...ZERO };
      addTokens(tb, tk);
      todayModelTokens.set(key, tb);
    }
  }
  s.modelTokens = Array.from(modelTokens.entries())
    .map(([model, tokens]) => ({ model, tokens }))
    .sort((a, b) => b.tokens.input - a.tokens.input);
  s.modelTokensToday = Array.from(todayModelTokens.entries())
    .map(([model, tokens]) => ({ model, tokens }))
    .sort((a, b) => b.tokens.input - a.tokens.input);
  // 标题兜底：无 session/title 时用首条消息截断
  if (!s.title && s.firstMsg) s.title = s.firstMsg.length > 40 ? s.firstMsg.slice(0, 40) + '…' : s.firstMsg;
  s.toolList = Object.entries(s.tools).sort((a, b) => b[1] - a[1]).slice(0, 3).map(([name, calls]) => ({ name, calls }));
  return s;
}

// 聚合所有 workspace 的会话统计
function scan() {
  const workspaces = [];
  const sessions = [];
  let total = { sessions: 0, turns: 0, steps: 0, toolCalls: 0, tokens: { ...ZERO }, tokensToday: { ...ZERO } };
  let latestTs = 0;
  let dirs;
  try { dirs = fs.readdirSync(SESSIONS_ROOT); } catch (e) { return { workspaces: [], sessions, total, latestTs, updatedAt: new Date().toISOString() }; }

  for (const wsDir of dirs) {
    const wsPath = path.join(SESSIONS_ROOT, wsDir);
    let st;
    try { st = fs.statSync(wsPath); } catch (e) { continue; }
    if (!st.isDirectory()) continue;
    let sids;
    try { sids = fs.readdirSync(wsPath); } catch (e) { continue; }
    const fallbackName = wsDir.replace(/^--|--$/g, '');
    const w = {
      name: fallbackName, // 首个会话真实 cwd basename 会覆盖（目录名是有损编码）
      sessions: 0, turns: 0, steps: 0, toolCalls: 0,
      tokens: { ...ZERO }, tokensToday: { ...ZERO }, latestTs: 0,
    };
    for (const sid of sids) {
      const logPath = path.join(wsPath, sid, 'session.jsonl.zstd');
      if (!fs.existsSync(logPath)) continue;
      const s = statLog(logPath, sid);
      if (!s) continue;
      const wsName = s.cwd ? path.basename(s.cwd) : fallbackName;
      if (w.sessions === 0 && wsName !== fallbackName) w.name = wsName;
      w.sessions++; w.turns += s.turns; w.steps += s.steps; w.toolCalls += s.toolCalls;
      addTokens(w.tokens, s.tokens);
      addTokens(w.tokensToday, s.tokensToday);
      if (s.lastTs > w.latestTs) w.latestTs = s.lastTs;
      sessions.push({ ...s, ws: wsName });
    }
    if (!w.sessions) continue;
    workspaces.push(w);
    total.sessions += w.sessions; total.turns += w.turns; total.steps += w.steps;
    total.toolCalls += w.toolCalls; addTokens(total.tokens, w.tokens);
    addTokens(total.tokensToday, w.tokensToday);
    if (w.latestTs > latestTs) latestTs = w.latestTs;
  }
  workspaces.sort((a, b) => b.latestTs - a.latestTs);
  // 会话卡按最近活动排序，截断到 MAX_SESSION_LIST
  sessions.sort((a, b) => b.lastTs - a.lastTs);
  // 全局模型分布（按 billed 输入降序），供大字悬浮 pop
  const modelMap = new Map();
  for (const s of sessions) {
    for (const mt of s.modelTokens || []) {
      const bucket = modelMap.get(mt.model) || { ...ZERO };
      addTokens(bucket, mt.tokens);
      modelMap.set(mt.model, bucket);
    }
  }
  const models = Array.from(modelMap.entries())
    .map(([model, tokens]) => ({ model, tokens }))
    .sort((a, b) => b.tokens.input - a.tokens.input);
  // 今日分模型分布（供 DSH 视图大字旁 ▦ 悬浮）
  const todayModelMap = new Map();
  for (const s of sessions) {
    for (const mt of s.modelTokensToday || []) {
      const bucket = todayModelMap.get(mt.model) || { ...ZERO };
      addTokens(bucket, mt.tokens);
      todayModelMap.set(mt.model, bucket);
    }
  }
  const todayModels = Array.from(todayModelMap.entries())
    .map(([model, tokens]) => ({ model, tokens }))
    .sort((a, b) => b.tokens.input - a.tokens.input);
  return {
    workspaces,
    sessions: sessions.slice(0, MAX_SESSION_LIST),
    total,
    models,
    todayModels,
    latestTs,
    updatedAt: new Date().toISOString(),
  };
}

function getDshUsage() {
  if (cache.payload && (Date.now() - cache.ts) / 1000 < CACHE_TTL) return cache.payload;
  cache = { ts: Date.now(), payload: scan() };
  return cache.payload;
}

module.exports = { getDshUsage, CACHE_TTL, _scan: scan };

File: src-tauri/src/data/dsh.rs
// DSH（DeepSeek Harness）用量统计 — 对齐 Electron src/data/dsh.js。
// 读 ~/.dsh/sessions 的 zstd 压缩会话日志（逐帧解压，对齐 JS 帧循环）。
// token 口径：usage 来自 assistant/chunk 与 assistant/message 的 usage，
// 同 turn/step 只取最后一份；输入 = uncachedInput + cacheRead + cacheWrite。
// 15s TTL 缓存，status() 路径零网络零阻塞。

use std::sync::{Mutex, OnceLock};

use serde_json::{json, Value};

use crate::data::sqlite::home_dir;
use crate::data::Cache;

const SESSIONS_ROOT: &str = ".dsh/sessions";
const CACHE_TTL_MS: u64 = 15_000;
const MAX_SESSION_LIST: usize = 10;
const ZSTD_MAGIC: [u8; 4] = [0x28, 0xb5, 0x2f, 0xfd];

fn cache() -> &'static Mutex<Cache> {
    static CACHE: OnceLock<Mutex<Cache>> = OnceLock::new();
    CACHE.get_or_init(|| Mutex::new(Cache::new()))
}

// 逐帧解压 zstd 流（每帧独立 frame，循环找 magic 边界解压拼接）。
// Rust 无 V8 GC 坑（Electron 版曾因 Buffer 进老生代累积数 GB OOM）。
fn decompress_zstd(buf: &[u8]) -> String {
    fn find_magic(hay: &[u8], from: usize) -> Option<usize> {
        (from..hay.len().saturating_sub(3)).find(|&i| hay[i..i + 4] == ZSTD_MAGIC)
    }
    let mut out = String::new();
    let mut pos = 0;
    while pos < buf.len() {
        let Some(idx) = find_magic(buf, pos) else { break };
        let next = find_magic(buf, idx + 4);
        let end = next.unwrap_or(buf.len());
        match zstd::stream::decode_all(std::io::Cursor::new(&buf[idx..end])) {
            Ok(decoded) => out.push_str(&String::from_utf8_lossy(&decoded)),
            Err(_) => break, // 尾帧不完整（正在写入）：丢弃
        }
        pos = end;
    }
    out
}

#[derive(Default, Clone)]
struct Tokens {
    input: i64,
    cache: i64,
    output: i64,
    reasoning: i64,
}

fn add_tokens(a: &mut Tokens, b: &Tokens) {
    a.input += b.input;
    a.cache += b.cache;
    a.output += b.output;
    a.reasoning += b.reasoning;
}

// 从 usage 记录取单步 token（输入 = billed 口径，对齐 GUI billedInputTokens）
fn tokens_from_usage(u: &Value) -> Option<Tokens> {
    let input = u.get("inputTokens").and_then(|v| v.as_i64());
    let out = u.get("outputTokens").and_then(|v| v.as_i64());
    if input.is_none() && out.is_none() {
        return None;
    }
    let cache = u.get("cacheReadTokens").and_then(|v| v.as_i64()).unwrap_or(0)
        + u.get("cacheWriteTokens").and_then(|v| v.as_i64()).unwrap_or(0);
    Some(Tokens {
        input: input.unwrap_or(0) + cache,
        cache,
        output: out.unwrap_or(0),
        reasoning: u.get("reasoningTokens").and_then(|v| v.as_i64()).unwrap_or(0),
    })
}

// 统计单个会话：精确 token + 标题/模型/工具/时间
fn stat_log(log_path: &std::path::Path, sid: &str) -> Option<Value> {
    let buf = std::fs::read(log_path).ok()?;
    let text = if log_path.to_string_lossy().ends_with(".zstd") {
        decompress_zstd(&buf)
    } else {
        String::from_utf8_lossy(&buf).to_string()
    };

    let mut s = json!({
        "id": sid, "title": "", "firstMsg": "", "model": "", "cwd": "",
        "turns": 0, "steps": 0, "toolCalls": 0,
        "createdAt": 0, "lastTs": 0,
    });
    // 同 turn/step 的 usage 只留最后一份（chunk 先行、message 收尾，后到替换不累加）
    let mut step_usage: std::collections::HashMap<String, (Tokens, String)> = std::collections::HashMap::new();
    let mut current_model = String::new();
    let mut tools: std::collections::HashMap<String, i64> = std::collections::HashMap::new();
    let mut turns = 0i64;
    let mut steps = 0i64;
    let mut tool_calls = 0i64;
    let mut created_at = 0i64;
    let mut last_ts = 0i64;

    for line in text.lines() {
        let Ok(j) = serde_json::from_str::<Value>(line) else { continue };
        let t = j.get("type").and_then(|v| v.as_str()).unwrap_or("");
        let d = j.get("data").cloned().unwrap_or(json!({}));
        match t {
            "session" => {
                created_at = j.get("createdAt").and_then(|v| v.as_i64()).unwrap_or(0);
                if let Some(cwd) = j.get("cwd").and_then(|v| v.as_str()) {
                    if !cwd.is_empty() {
                        s["cwd"] = json!(cwd);
                    }
                }
            }
            "turn/start" | "turn/end" => turns += 1,
            "step/start" | "step/end" => steps += 1,
            "tool/call" => {
                tool_calls += 1;
                if let Some(name) = d.get("name").and_then(|v| v.as_str()) {
                    *tools.entry(name.to_string()).or_insert(0) += 1;
                }
            }
            "request/context" => {
                if s.get("model").and_then(|v| v.as_str()).map(|m| m.is_empty()).unwrap_or(true) {
                    s["model"] = d
                        .get("model")
                        .or_else(|| d.get("provider"))
                        .cloned()
                        .unwrap_or(json!(""));
                }
                let provider = d.get("provider").and_then(|v| v.as_str()).unwrap_or("");
                let model = d.get("model").and_then(|v| v.as_str()).unwrap_or("");
                current_model = if provider.is_empty() { model.to_string() } else { format!("{provider}/{model}") };
            }
            "session/title" => {
                if let Some(title) = d.get("title").and_then(|v| v.as_str()) {
                    s["title"] = json!(title);
                }
            }
            "user/message" => {
                let first = s.get("firstMsg").and_then(|v| v.as_str()).map(|m| m.is_empty()).unwrap_or(true);
                if first {
                    if let Some(content) = d.get("content").and_then(|v| v.as_array()) {
                        if let Some(txt) = content.iter().find(|x| {
                            x.get("type").and_then(|v| v.as_str()) == Some("text")
                                && x.get("text").and_then(|v| v.as_str()).is_some()
                        }) {
                            s["firstMsg"] = txt.get("text").cloned().unwrap_or(json!(""));
                        }
                    }
                }
            }
            "assistant/chunk" => {
                // usage 类型 chunk：流的早期样本（token-meter 同源）
                if let Some(chunk) = d.get("chunk") {
                    if chunk.get("type").and_then(|v| v.as_str()) == Some("usage") {
                        if let Some(u) = chunk.get("usage") {
                            if let Some(tk) = tokens_from_usage(u) {
                                let key = format!("{}:{}", d.get("turn").and_then(|v| v.as_i64()).unwrap_or(0), d.get("step").and_then(|v| v.as_i64()).unwrap_or(0));
                                step_usage.insert(key, (tk, current_model.clone()));
                            }
                        }
                    }
                }
            }
            "assistant/message" => {
                // 组装消息的最终 usage：覆盖同 step 的 chunk 样本
                if let Some(u) = d.get("usage") {
                    if let Some(tk) = tokens_from_usage(u) {
                        let key = format!("{}:{}", d.get("turn").and_then(|v| v.as_i64()).unwrap_or(0), d.get("step").and_then(|v| v.as_i64()).unwrap_or(0));
                        step_usage.insert(key, (tk, current_model.clone()));
                    }
                }
            }
            _ => {}
        }
        let ts = j.get("time").and_then(|v| v.as_i64()).unwrap_or(0);
        if ts > last_ts {
            last_ts = ts;
        }
    }

    s["turns"] = json!(turns);
    s["steps"] = json!(steps);
    s["toolCalls"] = json!(tool_calls);
    s["createdAt"] = json!(created_at);
    s["lastTs"] = json!(last_ts);

    let mut total = Tokens::default();
    let mut model_tokens: std::collections::HashMap<String, Tokens> = std::collections::HashMap::new();
    for (tk, model) in step_usage.values() {
        add_tokens(&mut total, tk);
        let bucket = model_tokens.entry(if model.is_empty() { "unknown".into() } else { model.clone() }).or_default();
        add_tokens(bucket, tk);
    }
    s["tokens"] = json!({
        "input": total.input, "cache": total.cache,
        "output": total.output, "reasoning": total.reasoning,
    });
    let mut model_list: Vec<(String, Tokens)> = model_tokens.into_iter().collect();
    model_list.sort_by(|a, b| b.1.input.cmp(&a.1.input));
    s["modelTokens"] = json!(model_list
        .into_iter()
        .map(|(model, tokens)| json!({
            "model": model,
            "tokens": { "input": tokens.input, "cache": tokens.cache, "output": tokens.output, "reasoning": tokens.reasoning },
        }))
        .collect::<Vec<_>>());

    // 标题兜底：无 session/title 时用首条消息截断
    let title = s.get("title").and_then(|v| v.as_str()).unwrap_or("");
    let first = s.get("firstMsg").and_then(|v| v.as_str()).unwrap_or("");
    if title.is_empty() && !first.is_empty() {
        s["title"] = json!(if first.len() > 40 { format!("{}…", &first[..40]) } else { first.to_string() });
    }
    s["tools"] = json!(tools);
    let mut tool_list: Vec<(String, i64)> = tools.into_iter().collect();
    tool_list.sort_by(|a, b| b.1.cmp(&a.1));
    s["toolList"] = json!(tool_list
        .into_iter()
        .take(3)
        .map(|(name, calls)| json!({ "name": name, "calls": calls }))
        .collect::<Vec<_>>());
    Some(s)
}

// 聚合所有 workspace 的会话统计
fn scan() -> Value {
    let root = home_dir().join(SESSIONS_ROOT);
    let mut workspaces: Vec<Value> = Vec::new();
    let mut sessions: Vec<Value> = Vec::new();
    let mut total = json!({ "sessions": 0, "turns": 0, "steps": 0, "toolCalls": 0 });
    let mut total_tokens = Tokens::default();
    let mut latest_ts: i64 = 0;
    let mut global_models: std::collections::HashMap<String, Tokens> = std::collections::HashMap::new();

    let dirs = std::fs::read_dir(&root).map(|it| it.filter_map(|e| e.ok()).collect::<Vec<_>>()).unwrap_or_default();
    for entry in dirs {
        let ws_path = entry.path();
        if !ws_path.is_dir() {
            continue;
        }
        let ws_dir = entry.file_name().to_string_lossy().to_string();
        let fallback_name = ws_dir.trim_matches('-').to_string();
        let mut w = json!({
            "name": fallback_name,
            "sessions": 0, "turns": 0, "steps": 0, "toolCalls": 0,
            "latestTs": 0,
        });
        let mut w_tokens = Tokens::default();
        let mut w_sessions = 0i64;
        let mut w_turns = 0i64;
        let mut w_steps = 0i64;
        let mut w_tool_calls = 0i64;
        let mut w_latest_ts = 0i64;
        let mut w_name = fallback_name.clone();

        let sids = std::fs::read_dir(&ws_path).map(|it| it.filter_map(|e| e.ok()).collect::<Vec<_>>()).unwrap_or_default();
        for sid_entry in sids {
            let log_path = sid_entry.path().join("session.jsonl.zstd");
            if !log_path.exists() {
                continue;
            }
            let sid = sid_entry.file_name().to_string_lossy().to_string();
            let Some(s) = stat_log(&log_path, &sid) else { continue };
            // workspace 名用首个会话真实 cwd basename 覆盖（目录名是有损编码）
            if w_sessions == 0 {
                if let Some(cwd) = s.get("cwd").and_then(|v| v.as_str()) {
                    if let Some(base) = std::path::Path::new(cwd).file_name() {
                        w_name = base.to_string_lossy().to_string();
                    }
                }
            }
            w_sessions += 1;
            w_turns += s.get("turns").and_then(|v| v.as_i64()).unwrap_or(0);
            w_steps += s.get("steps").and_then(|v| v.as_i64()).unwrap_or(0);
            w_tool_calls += s.get("toolCalls").and_then(|v| v.as_i64()).unwrap_or(0);
            if let Some(tk) = s.get("tokens") {
                add_tokens(&mut w_tokens, &tokens_from_value(tk));
            }
            let st = s.get("lastTs").and_then(|v| v.as_i64()).unwrap_or(0);
            if st > w_latest_ts {
                w_latest_ts = st;
            }
            let mut s = s.clone();
            s["ws"] = json!(w_name.clone());
            sessions.push(s);
        }

        if w_sessions == 0 {
            continue;
        }
        w["name"] = json!(w_name);
        w["sessions"] = json!(w_sessions);
        w["turns"] = json!(w_turns);
        w["steps"] = json!(w_steps);
        w["toolCalls"] = json!(w_tool_calls);
        w["tokens"] = json!({
            "input": w_tokens.input, "cache": w_tokens.cache,
            "output": w_tokens.output, "reasoning": w_tokens.reasoning,
        });
        w["latestTs"] = json!(w_latest_ts);
        workspaces.push(w);

        total["sessions"] = json!(total["sessions"].as_i64().unwrap_or(0) + w_sessions);
        total["turns"] = json!(total["turns"].as_i64().unwrap_or(0) + w_turns);
        total["steps"] = json!(total["steps"].as_i64().unwrap_or(0) + w_steps);
        total["toolCalls"] = json!(total["toolCalls"].as_i64().unwrap_or(0) + w_tool_calls);
        add_tokens(&mut total_tokens, &w_tokens);
        if w_latest_ts > latest_ts {
            latest_ts = w_latest_ts;
        }
    }

    workspaces.sort_by(|a, b| b.get("latestTs").and_then(|v| v.as_i64()).cmp(&a.get("latestTs").and_then(|v| v.as_i64())));
    sessions.sort_by(|a, b| b.get("lastTs").and_then(|v| v.as_i64()).cmp(&a.get("lastTs").and_then(|v| v.as_i64())));
    // 全局模型分布（按 billed 输入降序），供大字悬浮 pop
    for s in &sessions {
        if let Some(mts) = s.get("modelTokens").and_then(|v| v.as_array()) {
            for mt in mts {
                let model = mt.get("model").and_then(|v| v.as_str()).unwrap_or("unknown").to_string();
                let bucket = global_models.entry(model).or_default();
                add_tokens(bucket, &tokens_from_value(mt.get("tokens").unwrap_or(&json!({}))));
            }
        }
    }
    let mut models: Vec<(String, Tokens)> = global_models.into_iter().collect();
    models.sort_by(|a, b| b.1.input.cmp(&a.1.input));

    total["tokens"] = json!({
        "input": total_tokens.input, "cache": total_tokens.cache,
        "output": total_tokens.output, "reasoning": total_tokens.reasoning,
    });

    json!({
        "workspaces": workspaces,
        "sessions": sessions.into_iter().take(MAX_SESSION_LIST).collect::<Vec<_>>(),
        "total": total,
        "models": models
            .into_iter()
            .map(|(model, tokens)| json!({
                "model": model,
                "tokens": { "input": tokens.input, "cache": tokens.cache, "output": tokens.output, "reasoning": tokens.reasoning },
            }))
            .collect::<Vec<_>>(),
        "latestTs": latest_ts,
        "updatedAt": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
    })
}

fn tokens_from_value(v: &Value) -> Tokens {
    Tokens {
        input: v.get("input").and_then(|x| x.as_i64()).unwrap_or(0),
        cache: v.get("cache").and_then(|x| x.as_i64()).unwrap_or(0),
        output: v.get("output").and_then(|x| x.as_i64()).unwrap_or(0),
        reasoning: v.get("reasoning").and_then(|x| x.as_i64()).unwrap_or(0),
    }
}

pub fn get_dsh_usage() -> Value {
    let mut guard = cache().lock().unwrap();
    if guard.fresh(CACHE_TTL_MS) {
        return guard.get();
    }
    let v = scan();
    guard.set(v.clone());
    v
}
